CyberArk/cyberark_modules/client/auth.py

```diff
@@ -0,0 +1,74 @@
+from datetime import datetime, timedelta
+from urllib.parse import urljoin
+
+import requests
+from requests.auth import AuthBase, HTTPBasicAuth
+from requests_ratelimiter import LimiterAdapter
+from urllib3.util.retry import Retry
+
+
+class CyberArkApiCredentials:
+    token_type: str
+    access_token: str
+    expires_at: datetime
+
+    @property
+    def authorization(self) -> str:
+        return f"{self.token_type.title()} {self.access_token}"
+
+
+class CyberArkApiAuthentication(AuthBase):
+    def __init__(
+        self,
+        auth_base_url: str,
+        application_id: str,
+        client_id: str,
+        client_secret: str,
+        api_key: str,
+        ratelimit_per_second: int = 10,
+    ) -> None:
+        self.__authorization_url = urljoin(auth_base_url, f"oauth2/token/{application_id}")
+        self.__client_id = client_id
+        self.__client_secret = client_secret
+        self.__api_key = api_key
+        self.__api_credentials: CyberArkApiCredentials | None = None
+
+        self.__http_session = requests.Session()
+        self.__http_session.auth = HTTPBasicAuth(username=self.__client_id, password=self.__client_secret)
+        self.__http_session.mount(
+            "https://",
+            LimiterAdapter(
+                per_second=ratelimit_per_second,
+                max_retries=Retry(
+                    total=5,
+                    backoff_factor=1,
+                ),
+            ),
+        )
+
+    def get_credentials(self) -> CyberArkApiCredentials:
+        current_dt = datetime.utcnow()
+
+        if self.__api_credentials is None or current_dt + timedelta(seconds=300) >= self.__api_credentials.expires_at:
+            response = self.__http_session.post(
+                url=self.__authorization_url,
+                data="grant_type=client_credentials&scope=isp.audit.events:read",
+                headers={"Content-Type": "application/x-www-form-urlencoded"},
+                timeout=60,
+            )
+            response.raise_for_status()
```

Comment:
Can you handle `Authentication error reponses` with status code 400?
Maybe you can create a specific Exception class, raise it here and intercept it in the connector.